frontend/packages/topology/src/elements/BaseEdge.ts

```diff
@@ -56,6 +59,20 @@ export default class BaseEdge<E extends EdgeModel = EdgeModel, D = any> extends
     this.target = target;
   }
 
+  getSourceAnchorNode(): Node | null {
+    if (!this.source) {
+      throw new Error(`Edge with ID '${this.getId()}' has no source.`);
+    }
+    return getClosestVisibleParent(this.source);
```

Comment:
I don't think the we should be involving visibility at this point.
We should only be looking at anchoring to a different source if a parent is collapsed. Therefore we should find the last collapsed parent in the hierarchy. Otherwise return `this.source`.

This would also mean that we can ensure a source anchor is always present `getSourceAnchorNode(): Node {`
It may not be visible, but always present. Otherwise we throw.